### PR TITLE
Add sequence variant for DATE

### PR DESCRIPTION
**Pull request by ngqtien**
As mentioned at issue https://github.com/prestodb/presto/issues/9681, the current return type for `SEQUENCE` function is `TIMESTAMP`, but we want to support `DATE` as well when we're using date type for start and stop arguments.

### CODE REVIEW

File: presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java [hunk 1]
```diff
@@ -1435,8 +1470,16 @@ private static SqlTimestamp sqlTimestamp(long millisUtc)
     private static SqlTimestamp sqlTimestamp(String dateString)
             throws ParseException
     {
-        DateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd");
+        DateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java]: This change should be a separate commit.

Comment (ngqtien) [reply presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java]: I don't think it's necessary because that function was added to test the `SEQUENCE` for `TIMESTAMP` return type support only. However, the tests are completely modified now to support `DATE` return type. Hence, I believe it should be put in the same commit, so people understand the context more clearly later on.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/SequenceFunction.java [hunk 2]
```diff
@@ -96,19 +110,53 @@ public static Block sequenceTimestampYearToMonth(
...
+        checkMaxEntryCondition(length);
 
         BlockBuilder blockBuilder = type.createBlockBuilder(new BlockBuilderStatus(), length);
         for (long i = 0, value = start; i < length; ++i, value += step) {
             type.writeLong(blockBuilder, value);
         }
         return blockBuilder.build();
     }
+
+    private static void checkCommonInvalidConditions(long start, long stop, long step)
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/SequenceFunction.java]: I can see the merit in extracting the `checkStates` (although I probably wouldn't have done this). But `checkCommonInvalidConditions` is a terrible name. Maybe call it `checkValidStep` instead?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/SequenceFunction.java [hunk 3]
```diff
@@ -96,19 +110,53 @@ public static Block sequenceTimestampYearToMonth(
...
+    {
+        checkCondition(step != 0, INVALID_FUNCTION_ARGUMENT, "step must not be zero");
+        checkCondition(step > 0 ? stop >= start : stop < start, INVALID_FUNCTION_ARGUMENT,
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/SequenceFunction.java]: The original formatting is not consistent with the preferred style. Write it as

```
checkCondition(
        step > 0 ? stop >= start : stop < start,
        INVALID_FUNCTION_ARGUMENT,
        "sequence stop value should be greater than or equal to start value if step is greater than zero otherwise stop should be less than start");
```

We prefer putting each argument (including the first one) on its own line.

File: presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java (hunk 1 above)
Comment (haozhun) [reply presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java]: @findepi, I have the same opinion as you at the beginning. But I realize that it wouldn't help in this case because every line that got touched in that pre commit will have to be touched again in the main commit anyways. I think it's fair as is.

@ngqtien, In general, we avoid having refactor and other change in the same file. Doing so makes reviewing this real changes much harder. But as I said in the previous paragraph, this is a special case.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/SequenceFunction.java (hunk 3 above)
Comment (haozhun) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/SequenceFunction.java]: @ngqtien Have a separate commit that change "<" to "<=".

File: presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java [hunk 4]
```diff
@@ -1296,45 +1299,77 @@ public void testSequence()
...
-        assertFunction("SEQUENCE(date '2016-04-12', date '2016-04-14', interval '1' day)", new ArrayType(TIMESTAMP),
-                ImmutableList.of(sqlTimestamp("2016-04-12"), sqlTimestamp("2016-04-13"), sqlTimestamp("2016-04-14")));
-        assertFunction("SEQUENCE(date '2016-04-14', date '2016-04-12', interval '-1' day)", new ArrayType(TIMESTAMP),
-                ImmutableList.of(sqlTimestamp("2016-04-14"), sqlTimestamp("2016-04-13"), sqlTimestamp("2016-04-12")));
-
-        assertFunction("SEQUENCE(date '2016-04-12', date '2016-04-16', interval '2' day)", new ArrayType(TIMESTAMP),
-                ImmutableList.of(sqlTimestamp("2016-04-12"), sqlTimestamp("2016-04-14"), sqlTimestamp("2016-04-16")));
-        assertFunction("SEQUENCE(date '2016-04-16', date '2016-04-12', interval '-2' day)", new ArrayType(TIMESTAMP),
-                ImmutableList.of(sqlTimestamp("2016-04-16"), sqlTimestamp("2016-04-14"), sqlTimestamp("2016-04-12")));
+        assertFunction("SEQUENCE(date '2016-04-12', date '2016-04-14', interval '1' day)", new ArrayType(DATE),
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java]: Put each argument on its own line

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/SequenceFunction.java [hunk 5]
```diff
@@ -157,9 +157,9 @@ private static void checkValidStep(long start, long stop, long step)
                 INVALID_FUNCTION_ARGUMENT,
                 "step must not be zero");
         checkCondition(
-                step > 0 ? stop >= start : stop < start,
+                step > 0 ? stop >= start : stop <= start,
                 INVALID_FUNCTION_ARGUMENT,
-                "sequence stop value should be greater than or equal to start value if step is greater than zero otherwise stop should be less than start");
+                "sequence stop value should be greater than or equal to start value if step is greater than zero otherwise stop should be less than or equal start");
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/SequenceFunction.java]: `or equal to`

**Comment by haozhun** [comment]
The email address in the Author field of your commits is incorrect. (It says "--help".)

Please fix it and make sure the email address is linked to your github account.

    git config --global user.name "Your Name"
    git config --global user.email you@example.com

After doing this, you may fix the identity used for this commit with:

    git commit --amend --reset-author

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/SequenceFunction.java [hunk 6]
```diff
@@ -96,19 +112,61 @@ public static Block sequenceTimestampYearToMonth(
...
+                step != 0,
+                INVALID_FUNCTION_ARGUMENT,
+                "step must not be zero");
+        checkCondition(
+                step > 0 ? stop >= start : stop <= start,
+                INVALID_FUNCTION_ARGUMENT,
+                "sequence stop value should be greater than or equal to start value if step is greater than zero otherwise stop should be less than or equal start");
+    }
+
+    private static void checkMaxEntryCondition(int length)
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/SequenceFunction.java]: rename to `checkMaxEntry`

File: presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java [hunk 7]
```diff
@@ -1296,45 +1299,143 @@ public void testSequence()
...
+                new ArrayType(TIMESTAMP),
+                ImmutableList.of(sqlTimestamp("2016-04-16 01:00:10"), sqlTimestamp("2016-04-15 06:00:10"), sqlTimestamp("2016-04-14 11:00:10")));
 
         // failure modes
-        assertInvalidFunction("SEQUENCE(date '2016-04-12', date '2016-04-14', interval '-1' day)", INVALID_FUNCTION_ARGUMENT);
-        assertInvalidFunction("SEQUENCE(date '2016-04-14', date '2016-04-12', interval '1' day)", INVALID_FUNCTION_ARGUMENT);
-        assertInvalidFunction("SEQUENCE(date '2000-04-14', date '2030-04-12', interval '1' day)", INVALID_FUNCTION_ARGUMENT);
+        assertInvalidFunction(
+                "SEQUENCE(date '2016-04-12', date '2016-04-14', interval '-1' day)",
+                INVALID_FUNCTION_ARGUMENT);
```
Comment (electrum) [inline presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java]: Can you change all of these to have the expected error message rather than `INVALID_FUNCTION_ARGUMENT`?

**Comment by electrum** [comment]
Should we also add a two argument version `sequence(start, end)` that increments or decrements by one day, similar to the two argument version for integers? That would seem to be the dominant use case, so having a short-hand would be convenient.

File: presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java [hunk 8]
```diff
@@ -1269,30 +1269,91 @@ public void testRepeat()
...
     public void testSequence()
+            throws ParseException
     {
         // defaults to a step of 1
-        assertFunction("SEQUENCE(1, 5)", new ArrayType(BIGINT), ImmutableList.of(1L, 2L, 3L, 4L, 5L));
-        assertFunction("SEQUENCE(-10, -5)", new ArrayType(BIGINT), ImmutableList.of(-10L, -9L, -8L, -7L, -6L, -5L));
-        assertFunction("SEQUENCE(-5, 2)", new ArrayType(BIGINT), ImmutableList.of(-5L, -4L, -3L, -2L, -1L, 0L, 1L, 2L));
-        assertFunction("SEQUENCE(2, 2)", new ArrayType(BIGINT), ImmutableList.of(2L));
+        assertFunction(
+                "SEQUENCE(1, 5)",
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java]: I don't understand why you did the reformatting here. Can you undo it please?

**Comment by haozhun** [comment]
@ngqtien The commit message for the first two commit has space for improvement. I change them to "Add date variant of sequence function" and "Fix argument validation in sequence function". I happen to be working on something that potentially conflict with this. As a result, I went ahead and merged the first 2 commits.

When you rebase and "push -f", you will only see your last commit in the pull request. That's expect and please don't be worried.

File: presto-docs/src/main/sphinx/functions/array.rst [hunk 9]
```diff
@@ -133,6 +133,11 @@ Array Functions
     Generate a sequence of integers from ``start`` to ``stop``, incrementing
     by ``1`` if ``start`` is less than or equal to ``stop``, otherwise ``-1``.
 
+.. function:: sequence(start, stop) -> array<date>
```
Comment (electrum) [inline presto-docs/src/main/sphinx/functions/array.rst]: Move this down to be above `sequence(start, stop, step) -> array<date>` variant

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/SequenceFunction.java [hunk 10]
```diff
@@ -64,6 +64,15 @@ public static Block sequenceDefaultStep(
         return fixedWidthSequence(start, stop, stop >= start ? 1 : -1, BIGINT);
     }
 
+    @ScalarFunction("sequence")
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/SequenceFunction.java]: Also move this down

**Comment by ngqtien** [comment]
@electrum I re-ordered other functions as well to match with the order from `array.rst`

**Comment by haozhun** [comment]
Merged #10077. Thank you.
